### PR TITLE
Migration and countersigning content updates, 11ty site, take 2

### PR DESCRIPTION
**Pull request by pdaoust**
Trying this again, because my attempt to merge an old mkdocs branch into the 11ty branch didn't work so well.

Sorry to all who like clean separation of concerns in PRs -- this one blends Core Concepts and Glossary work on both countersigning and migration. It supersedes https://github.com/holochain/docs-pages/pull/313 because (a) it incorporates some of the feedback from that PR and (b) it conforms to the new 11ty project structure.

@thedavidmeister requesting your review for correctness -- and clarity too if you'd like, but don't sweat it if not.

### CODE REVIEW

**Comment by pdaoust** [comment]
Converting to a draft because it doesn't have images. @marycamacho @MatthewNichols if you'd rather get the content out in its current form, go ahead and un-draft it once it's been reviewed. Hoping I can create some images before Thurs, but not sure yet.

**Comment by marycamacho** [comment]
@pdaoust Let's not try to merge this into this conversion branch - let's wait until after the deploy of the new design, have all the correctness reviews and images ready and then merge into main.

**Comment by pdaoust** [comment]
Renamed this PR to 'countersigning' only and reverted all the migration stuff, cuz it's not clear to me how migration is supposed to work, and Meister also says that any content would have to be so dependent on the appropriate strategy for a hApp. I've touched on strategies in the page, but it's not well-connected to anything concrete.

**Comment by pdaoust** [comment]
@jost-s @marycamacho I'd love your review on this, if you have the bandwidth -- there's a staggering amount of changes, I must confess, but they correct a lot of obsolete information in the core concepts and glossary, and I think they make the core concepts much easier to read.

Looking for feedback on the value these changes do or don't create (although of course if you see a mistake, please give feedback on that too!)

File: src/pages/references/glossary.md [hunk 1]
```diff
@@ -598,56 +776,79 @@ The programming language currently used to build Holochain Core and [DNAs](#dna)
...
+A modification of application state. In Holochain, all state transitions are initially created as [records](#record) in an [agent's](#agent) [source chain](#source-chain) that represent the [actions](#action) of [creating, updating, and deleting](#create-read-update-delete-crud) data and metadata, as well as of system-level actions such as [capability grants](#capability-grant). A state transition then yields one or more [operations](#dht-operation) that are then [published](#publish) to the [DHT](#distributed-hash-table-dht), that is, they are sent to the appropriate [validation authorities](#validation-authority), who then apply those operations to their own DHT [shard](#sharding), which causes a state transition for the [base](#base) to which the operation applies.
```
Comment (jost-s) [inline src/pages/references/glossary.md:810]:
```suggestion
A modification of application state. In Holochain, all state transitions are initially created as [records](#record) in an [agent's](#agent) [source chain](#source-chain) that represent the [actions](#action) of [creating, updating, and deleting](#create-read-update-delete-crud) data and metadata, as well as of system-level actions such as [capability grants](#capability-grant). A state transition further yields one or more [operations](#dht-operation) that are then [published](#publish) to the [DHT](#distributed-hash-table-dht), that is, they are sent to the appropriate [validation authorities](#validation-authority), who then apply those operations to their own DHT [shard](#sharding), which causes a state transition for the [base](#base) to which the operation applies.
```

File: src/pages/references/glossary.md [hunk 2]
```diff
@@ -404,95 +500,142 @@ See [Holochain host API](#holochain-host-api).
...
 
-#### Init complete record
+#### Init complete action
 
-A [record](#record) that Holochain automatically writes to an [agent](#agent)'s [source chain](#source-chain) to indicate that all of a [DNA](#dna)'s [init callbacks](#init-callback) have successfully run and their [cell](#cell) is ready to use.
+A [record](#record) that Holochain automatically writes to an [agent's](#agent) [source chain](#source-chain) to indicate that the [init callbacks](#init-callback) in all of a [DNA's](#dna) [zomes](#zome) have successfully run and the [cell](#cell) is ready to use.
+
+#### Integrity zome
+
+A [zome](#zome) that defines a data schema. It does this through three specially named [zome functions](#zome-function):
```
Comment (jost-s) [inline src/pages/references/glossary.md:515]: Entry types and link types merely are enums in Rust. I was about to add `domain logic zome functions` to the coordinator zome above, but then thought that zome functions are exclusive to coordinator zomes and don't need that more specific descriptor. I wouldn't consider the validation function in integrity zomes a "zome function".
```suggestion
A [zome](#zome) that defines a data schema. It does this through three specially named constructs:
```

File: src/pages/references/glossary.md [hunk 3]
```diff
@@ -404,95 +500,142 @@ See [Holochain host API](#holochain-host-api).
...
 #### Init callback
 
-A function in a [DNA](#dna) that the [nucleus](#nucleus) calls when an [agent](#agent) starts a [cell](#cell) for the first time. This can be used to set up initial [source chain](#source-chain) [#state](#state-transition), etc.
+A function in a [DNA](#dna) that the [conductor](#conductor) calls when an [agent](#agent) starts a [cell](#cell) for the first time, and after they have joined the DNA's [network](#network). This can be used to set up initial [source chain](#source-chain) [#state](#state-transition), etc.
 
-#### Init complete record
+#### Init complete action
 
-A [record](#record) that Holochain automatically writes to an [agent](#agent)'s [source chain](#source-chain) to indicate that all of a [DNA](#dna)'s [init callbacks](#init-callback) have successfully run and their [cell](#cell) is ready to use.
+A [record](#record) that Holochain automatically writes to an [agent's](#agent) [source chain](#source-chain) to indicate that the [init callbacks](#init-callback) in all of a [DNA's](#dna) [zomes](#zome) have successfully run and the [cell](#cell) is ready to use.
```
Comment (jost-s) [inline src/pages/references/glossary.md:511]: Strictly the cell can be used before having been lazily initialized as part of the first zome call.
```suggestion
An [action](#action) that Holochain automatically writes to an [agent's](#agent) [source chain](#source-chain) to indicate that the [init callbacks](#init-callback) in all of a [DNA's](#dna) [zomes](#zome) have successfully run..
```

File: src/pages/references/glossary.md [hunk 4]
```diff
@@ -404,95 +500,142 @@ See [Holochain host API](#holochain-host-api).
...
+A property of Holochain's [validating DHT](#validating-dht), whereby healthy [nodes](#node) detect invalid data, share proof of corruption among their peers via [warrants](#warrant), and take defensive action against the corrupt nodes that [authored](#author) the data by blocking network communication with them. While each agent is individually free to interact with a warranted peer, most agents will refuse to interact or gossip with them. The cumulative effect is a collective exclusion of the corrupt nodes (see [mutual sovereignty](#mutual-sovereignty)).
 
 #### Init callback
 
-A function in a [DNA](#dna) that the [nucleus](#nucleus) calls when an [agent](#agent) starts a [cell](#cell) for the first time. This can be used to set up initial [source chain](#source-chain) [#state](#state-transition), etc.
+A function in a [DNA](#dna) that the [conductor](#conductor) calls when an [agent](#agent) starts a [cell](#cell) for the first time, and after they have joined the DNA's [network](#network). This can be used to set up initial [source chain](#source-chain) [#state](#state-transition), etc.
```
Comment (jost-s) [inline src/pages/references/glossary.md:507]:
```suggestion
A function in a [DNA](#dna) that the [conductor](#conductor) executes when an [agent](#agent) calls a [cell](#cell) for the first time, and after they have joined the DNA's [network](#network). This can be used to set up initial [source chain](#source-chain) [#state](#state-transition), etc.
```

**Review by jost-s — CHANGES_REQUESTED** [review]
Wow, this is huge! I will read through this again. I may well have missed things.

I see more clarity and precision, many updates to match the current codebase. There's so much technical information useful for hApp devs that I will add lots of links from the Rust docs to specific places here once it's merged.

**Comment by jost-s** [comment]
I'm wondering in how far this body of work could/should be combined with the Holochain specs that are being added to the repo by @zippy. With the technical depth that this glossary and the chapters are achieving, it might make sense, to not have to maintain two different documents.

**Review by thedavidmeister — APPROVED** [review]
i went over the countersigning stuff, looks good

File: src/pages/concepts/2_application_architecture.md [hunk 5]
```diff
@@ -30,15 +30,15 @@ These data integrity rules create a membrane between a participant and her peers
...
 
 ![One participant's copy of the hApp. A membrane surrounds her hApp. On her device, three clients are successfully calling the hApp's functions while a malware is rejected. From the network, two peers try to call her hApp's functions; one succeds while another fails.](/assets/img/concepts/2.3-process-membrane.png){.sz80p} {.center}
 
 ## Layers of the application stack
 
-Holochain handles a lot of things for you, keeping your workload minimal. That’s why we call it a framework. You rely on Holochain for things like data persistence and a peer-to-peer networking and communication layer, and get on with building your application and business logic.
+Holochain handles a lot of things for you, keeping your development workload minimal and mostly focused on the problems you need to salf. That's why we call it a framework. You rely on Holochain for things like data persistence and a peer-to-peer networking layer, and get on with building your application and business logic.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/2_application_architecture.md:39]: one solf, two solves -- just like wolf/wolves 🤦 thank you for catching that

File: src/pages/concepts/2_application_architecture.md [hunk 6]
```diff
@@ -52,58 +52,61 @@ All clear? Don't worry; it'll make more sense as we dig in.
...
 
 Each cell stores its own agent history and is a member of an isolated network along with other cells that share the same DNA. This is useful for creating separate spaces inside a single hApp --- individual project workspaces or private chat channels, for example.
 
-When a client calls a function in a hApp, it specifies the **cell ID**, which is the hash of the DNA plus the agent ID, along with the zome name, function name, and function input payload.
+As a bundle of running code, a cell exposes the functions you write as its public API, which can be accessed by other cells within the same hApp, clients on the same machine, or other cells on the network that the cell belongs to. When a client or other cell calls a function in a hApp, it specifies the **cell ID**, which is the hash of the DNA plus the agent ID, along with the zome name, function name, and function input payload. The developer can give a participant the ability to control access to their cell's API via [capabilities](../8_calls_capabilities/).
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/2_application_architecture.md:81]: ah, interesting, shows my language wasn't clear -- the participant in this case would be the cell owner. I'll add a suggestion of my own to fix that.

Comment (pdaoust) [inline src/pages/concepts/2_application_architecture.md:81]:
```suggestion
As a bundle of running code, a cell exposes the functions you write as its public API, which can be accessed by other cells within the same hApp, clients on the same machine, or other cells on the network that the cell belongs to. When a client or other cell calls a function in a hApp, it specifies the **cell ID**, which is the hash of the DNA plus the agent ID, along with the zome name, function name, and function input payload. The developer can give a cell owner the ability to control access to their cell's API via [capabilities](../8_calls_capabilities/).
```

File: src/pages/concepts/2_application_architecture.md [hunk 7]
```diff
@@ -52,58 +52,61 @@ All clear? Don't worry; it'll make more sense as we dig in.
...
+Consider source code and configuration changes carefully, because each modification to an integrity zome will create a new DNA with new cells interacting in a separate network. This may require some sort of migration strategy to move or access data between old and new cells.
 
-All of these functions are run from the perspective of the individual participant. When a client on Alice's computer calls a function that writes data, it calls that function in her own cell, writing data to her personal store. Unlike with cloud and blockchain, there is no objective, global-level actor. Things happen only when someone causes them to happen.
+Coordinator zomes aren't subject to the same strict rules, and can be swapped for new ones in a running cell. Care must still be taken, however, because each participant is free to swap coordinator zomes as they like (or more realistically, as the hApp developer gives them the power to do), so an entire network of cells isn't guaranteed to be running the same coordinator code.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/2_application_architecture.md:106]: ah, didn't know that. Thank you.

Comment (pdaoust) [inline src/pages/concepts/2_application_architecture.md:106]:
```suggestion
Coordinator zomes aren't subject to the same strict rules, and can be swapped for new ones in a running cell. Care must still be taken, however, because each participant is free to swap coordinator zomes as they like (or more realistically, as the hApp developer gives them the power to do), so an entire network of cells isn't guaranteed to be running the same coordinator code. (Note: While the functionality exists at the low level, there are no API methods to make this happen yet.)
```

Comment (pdaoust) [reply src/pages/concepts/2_application_architecture.md:106]: @jost-s is this true?

File: src/pages/concepts/3_source_chain.md [hunk 8]
```diff
@@ -47,65 +48,78 @@ All your key pairs are stored in an encrypted, password-protected key manager on
...
+
+## Concurrency, atomic commits, and source chain lifecycle events
+
+Multiple zome function calls can be made at one time, and each of them can read or write data as needed. The source chain state that the function sees is a snapshot from the beginning of function execution, however, so it won't see writes made by the other call. And if two calls try to write to the agent's source chain at once, the first one to finish will succeed and the second will fail, telling the caller that the source chain top has moved. The caller can then give up or try again until the call succeeds.
+
+If the ordering of records doesn't matter, the function can be written to use **relaxed chain ordering** -- that is, if it fails to write, the conductor will try the write again, adding the new records onto the end of the updated chain. (This is similar to [rebasing](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) in Git.)
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/3_source_chain.md:81]: Love the attention to detail. This is the sort of feedback I crave.

File: src/pages/concepts/4_dht.md [hunk 9]
```diff
@@ -17,126 +17,145 @@ Agents share records of their actions, including any data meant to be shared wit
...
-![](/assets/img/concepts/4.1-network.png)
+![](/assets/img/concepts/4.1-network.png){.sz80p} {.center}
 
 In a Holochain network, you share your source chain actions and public entries with a random selection of your peers, who witness, validate, and hold copies of them.
 
-![](/assets/img/concepts/4.2-public-entry-commit.png)
+![](/assets/img/concepts/4.2-public-entry-commit.png){.sz80p} {.center}
 
-When you commit a record with private data, the entry data stays in your source chain, but its action is still shared.
+When you commit a record with private data, the entry data stays in your source chain, but its action header is still shared.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/4_dht.md:43]: the language of 'action' is really awkward to work with. Here I'm adopting a term I made up, 'action header', to describe the actual bytes of the action data structure. Does that work, or is it just confusing?

File: src/pages/concepts/5_links_anchors.md [hunk 10]
```diff
@@ -35,74 +35,84 @@ So how do you discover the address of the data you're looking for?
...
 
-We recommend **anchors** as a useful pattern for creating starting points on the DHT’s graph database. Anchors are just small chunks of data, usually strings, that serve as things to attach links to. They can be easy for humans to share and type, such as usernames and hashtags, or they can be baked into the application as constants. The important thing is that they're simple enough that anyone can generate their address and retrieve links attached to them without knowing the address beforehand.
+We recommend **anchors** as a useful pattern for creating starting points on the DHT's graph database. An anchor is just a small chunk of data on the DHT, usually a string, that serves as bases to attach a lot of links to. An anchor's value can be easy for humans to share and type, such as a username or hashtag, or it can be baked into the application as a string constant. The important thing is that anchors should be easy to locate.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/5_links_anchors.md:57]: that's the part I always struggle to describe. It's obvious to me -- you don't have to go looking for it, cuz you already know what it is. That could mean that it's a constant or something simple that you type in. Suggestions welcome 🙏🏼

Comment (pdaoust) [inline src/pages/concepts/5_links_anchors.md:57]:
```suggestion
We recommend **anchors** as a useful pattern for creating starting points on the DHT's graph database. An anchor is just a small chunk of data on the DHT, usually a string, that serves as a basis to attach a lot of links to. An anchor's value can be easy for humans to share and type, such as a username or hashtag, or it can be baked into the application as a string constant. The important thing is that anchors should be easy to locate.
```

File: src/pages/concepts/7_validation.md [hunk 11]
```diff
@@ -3,208 +3,229 @@ title: "Validation: Assuring Data Integrity"
...
+Some entries can be computationally expensive to validate. In a currency app, for example, the validity of a transaction depends on the account balances of both transacting parties, which is the sum of all their prior transactions. The validity of each of those transactions depends on the account balance at the time, plus the validity of the account balance of the people they transacted with, and so on and so on. The data is deeply interconnected; you don't want to wait while the coffee shop's payment terminal interrogates half the town's economic history when you're just trying to buy a cup of coffee and you're already to work.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/7_validation.md:40]: ha, whoops, it was the 'late' one

Comment (pdaoust) [inline src/pages/concepts/7_validation.md:40]:
```suggestion
Some entries can be computationally expensive to validate. In a currency app, for example, the validity of a transaction depends on the account balances of both transacting parties, which is the sum of all their prior transactions. The validity of each of those transactions depends on the account balance at the time, plus the validity of the account balance of the people they transacted with, and so on and so on. The data is deeply interconnected; you don't want to wait while the coffee shop's payment terminal interrogates half the town's economic history when you're just trying to buy a cup of coffee and you're already late for work.
```

File: src/pages/concepts/7_validation.md [hunk 12]
```diff
@@ -3,208 +3,229 @@ title: "Validation: Assuring Data Integrity"
...
+The DHT offers a shortcut --- it remembers the validation results of existing entries. You can ask the validation authorities for the parties' previous transactions if they detected any problems. You can assume that they have done the same thing for the transaction prior to those and so on. As long as you trust a decent number of your peers to be playing by the rules, the validation results attached to the most recent entry 'proves' the validity of all the entries before it.
 
-You can also consult the agent activity authority for any peer you're unsure of. They hold a copy of any past evidence of malicious activity, along with a record of all of the peer's source chain actions that show whether they've tried to change their history.
+For evidence of attempts to You can also consult the agent activity authority for any peer you're unsure of. They may or hold a copy of any past evidence of malicious activity, along with a record of all of the peer's source chain actions that show whether they've tried to change their history.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/7_validation.md:44]: what the heck

Comment (pdaoust) [inline src/pages/concepts/7_validation.md:44]:
```suggestion
For evidence of attempts to fork a chain, you can also consult the agent activity authority for any peer you're unsure of. They hold a record of all of the peer's source chain actions that show whether they've tried to change their history.
```

It seems that agent activities will probably not act as clearinghouses for warrants, so I'll only make claims about fork attempts.

File: src/pages/concepts/7_validation.md [hunk 13]
```diff
@@ -3,208 +3,229 @@ title: "Validation: Assuring Data Integrity"
...
+When you **commit a record**, your Holochain conductor is responsible for making sure you're playing by the rules. This protects you from publishing any invalid data that could make you look like a bad actor.
 
 #### Valid entry
 
 ::: coreconcepts-storysequence
-1. ![](/assets/img/concepts/7.2-commit.png)
-Alice calls the `publish_word` zome function with the string `"eggplant"`. The function commits that word to her source chain. The conductor ‘stages’ the commit in the function’s scratch space and returns the creation action’s record hash to the `publish_word` function. The function continues executing and passes a return value back to the conductor, which holds onto it for now.
+![](/assets/img/concepts/7.2-commit.png){.sz80p} {.center}
+
+Alice calls the `publish_word` zome function with the string `"eggplant"`. The function commits that word to her source chain. The conductor 'stages' the commit in the function's scratch space and returns the creation action's record hash to the `publish_word` function. The function continues executing and passes a return value back to the conductor, which holds onto it for now.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/7_validation.md:86]: You used to be able to define validation functions like `validate_create_word`, which would follow a sort of cascade of specificity -- that is, for the act of publishing a `word` entry, the `validate` function (which validates all actions) would be called first, then `validate_create` (which validates only `CreateEntry` actions) next, followed by `validate_create_word` (which validates only `CreateEntry` actions that operate on `word`s). That complexity has disappeared, along with the idea of validating actions -- now it's ops that are validated.

tl;dr: `publish_word` should still be the name of the zome function.

File: src/pages/concepts/7_validation.md [hunk 14]
```diff
@@ -3,208 +3,229 @@ title: "Validation: Assuring Data Integrity"
...
+## Final note: genesis self-check
 
-If an agent is committing a record to their source chain that depends on DHT data, it's **their job to make sure those references exist at commit time and explicitly reference them**. The validation function doesn't need to revalidate those dependencies, though; Holochain will use its own heuristics to determine the trustworthiness of other validators' claims when it retrieves the data.
+There is one more pseudo-validation function you should know about: the **genesis self-check**. As with validation functions, you define this callback in your integrity zomes. Its job is to 'pre-validate' an agent's membrane proof before she joins a network, to prevent her from accidentally committing a membrane proof that woud bar her from the network forever.
```
Comment (pdaoust) [inline src/pages/concepts/7_validation.md:201]:
```suggestion
There is one more pseudo-validation function you should know about: the **genesis self-check**. As with validation functions, you define this callback in your integrity zomes. Its job is to 'pre-validate' an agent's membrane proof before she joins a network, to prevent her from accidentally committing a membrane proof that would bar her from participating in the network forever.
```

File: src/pages/concepts/8_calls_capabilities.md [hunk 15]
```diff
@@ -15,124 +15,122 @@ Application components can **call a DNA's functions**. On one agent’s device,
...
 
-An agent makes things happen in their cell by calling one of its public functions through the **app interface**, which is a WebSocket port that the conductor makes available on the agent's device. The thing making the calls is a client of some sort --- a GUI, a shell script, a long-running service, anything that can speak WebSocket. The important thing to remember is that, because the conductor only exposes the app interface on the local machine, the UI has to live on the local machine. This helps discourage anyone from impersonating the owner of the cell.
+An agent makes things happen in her cell by calling one of its public functions through the **app interface**, which is a WebSocket port that the conductor makes available on the agent's device. The thing making the calls is a client of some sort --- a GUI, a shell script, a long-running service, anything that can speak WebSocket. The important thing to remember is that, because the conductor only exposes the app interface on the local machine, the client has to live on the local machine too. This protects the cell from being accessed by anyone other than its owner.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/8_calls_capabilities.md:31]: yeah, thinking also about Flux and how it proxies the WebSocket over the LAN. Those are some interesting use cases, and I don't want to make it seem like they're impossible.

Comment (pdaoust) [inline src/pages/concepts/8_calls_capabilities.md:31]:
```suggestion
An agent makes things happen in her cell by calling one of its public functions through the **app interface**, which is a WebSocket port that the conductor makes available on the agent's device. The thing making the calls is a client of some sort --- a GUI, a shell script, a long-running service, anything that can speak WebSocket. The important thing to remember is that the conductor only exposes the app interface on the local machine, which protects the cell from being accessed by anyone other than its owner. It's possible to expose that interface to the world using an HTTP proxy, but care should be taken to secure it --- primarily by helping the owner of the hApp understand the risk and make wise choices about whether to grant a capability to a given client.
```

File: src/pages/concepts/9_signals.md [hunk 16]
```diff
@@ -12,52 +12,51 @@ A DNA usually only receives function calls from the outside world and returns a
 1. [When signals are useful](#when-are-signals-useful)
 2. [Where you can send a signal to](#local-and-remote-signals)
 3. [What signals should not be used for](#when-not-to-use-signals)
+4. [What the future holds](#the-future-pub-sub)
 
 ### <i class="far fa-atom"></i> Why it matters
 
-Signals avoid the need for a client to regularly poll a zome function to retrieve new data, making UIs much more responsive and performant. They can also trigger automatic actions in a cell or client without needing human intervention.
+Signals eliminate the need for a client to regularly poll a zome function to retrieve new data, making UIs much more responsive and performant. They can also trigger automatic actions in a cell or client without needing human intervention.
```
Comment (pdaoust) [inline src/pages/concepts/9_signals.md:19]:
```suggestion
Signals reduce the need for a client to regularly poll a zome function to retrieve new data, making UIs much more responsive and performant. They can also trigger automatic actions in a cell or client without needing human intervention.
```

File: src/pages/get-building/install-advanced.md [hunk 17]
```diff
@@ -215,7 +215,7 @@ It is important that these remain consistent, so you can get your work done with
 
 The main Nix tool used in Holochain development workflows is `nix develop`, a program that overlays a new Bash environment and set of tools on top of your existing shell environment.
 
-The full suite of Nix tooling is broad and deep. There’s even a dedicated programming language, called [Nix expressions](https://nixos.org/manual/nix/stable/#functional-package-language). Learn more with the [NixOS Wiki](https://nixos.wiki) or the [Pills](https://nixos.org/nixos/nix-pills/) Tutorial. The [NixOS community chat on matrix.to](https://matrix.to/#/#community:nixos.org) is aactive and helpful.
+The full suite of Nix tooling is broad and deep. There's even a dedicated programming language, called [Nix expressions](https://nixos.org/manual/nix/stable/#functional-package-language). Learn more with the [NixOS Wiki](https://nixos.wiki) or the [Pills](https://nixos.org/nixos/nix-pills/) Tutorial. The [NixOS community chat on matrix.to](https://matrix.to/#/#community:nixos.org) is aactive and helpful.
```
Comment (pdaoust) [inline src/pages/get-building/install-advanced.md:218]:
```suggestion
The full suite of Nix tooling is broad and deep. There's even a dedicated programming language, called [Nix expressions](https://nixos.org/manual/nix/stable/#functional-package-language). Learn more with the [NixOS Wiki](https://nixos.wiki) or the [Pills](https://nixos.org/nixos/nix-pills/) Tutorial. The [NixOS community chat on matrix.to](https://matrix.to/#/#community:nixos.org) is active and helpful.
```

File: src/pages/references/glossary.md [hunk 18]
```diff
@@ -28,156 +28,207 @@ The power of an [agent](#agent) to act in their environment.
...
+A highly [centralized](#centralization) [distributed system](#distributed-system) architecture in which certain privileged nodes are responsible for most of the processing, storage, and decision-making. Client/server systems typically give low [agency](#agency) to end-users, due to the fact that the privileged nodes mediate most user interactions.
 
-#### Clone (DNA)
+#### Cloning
 
-The act of modifying the properties of [DNA](#dna) at installation time in order to change its hash and create an isolated [network](#network) and [DHT](#distributed-hash-table-dht). This allows users to enjoy a private space using existing rules without creating a DNA from scratch.
+The act of modifying one or more [DNA modifiers](#dna-modifiers) at DNA installation time in order to change its [hash](#dna-hash) and optionally its behavior via [DNA properties](#dna-properties), thus creating an isolated [network](#network) and [DHT](#distributed-hash-table-dht) for that clone. This allows all [agents](#agent) who have applied the same modifications to enjoy a private shared space using existing rules without creating a DNA from scratch.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/references/glossary.md:188]: you also made the sentence work way better.

File: src/pages/references/glossary.md (hunk 4 above)
Comment (pdaoust) [reply src/pages/references/glossary.md:507]: Ah yes, this is true. I'd elided the fact that it didn't run until first zome call, but that's probably important lifecycle timing to know about.

File: src/pages/references/glossary.md (hunk 3 above)
Comment (pdaoust) [inline src/pages/references/glossary.md:511]:
```suggestion
An [action](#action) that Holochain automatically writes to an [agent's](#agent) [source chain](#source-chain) to indicate that the [init callbacks](#init-callback) in all of a [DNA's](#dna) [zomes](#zome) have successfully run.
```

There was an extra period at the end of the suggestion; otherwise I would've accepted it ;)

File: src/pages/references/glossary.md (hunk 2 above)
Comment (pdaoust) [reply src/pages/references/glossary.md:515]: oh, I thought the conductor retrieved them at installation time by calling an `entry_types` function?

File: src/pages/references/glossary.md [hunk 19]
```diff
@@ -404,95 +500,142 @@ See [Holochain host API](#holochain-host-api).
...
 
 #### Logical monotonicity
 
-The property of a set of facts whereby the truth of prior facts are never negated by the addition of later facts. [CALM](#consistency-as-logical-monotonicity-calm-theorem) relies on functions that exhibit this property. For example, Holochain's [source chain](#source-chain), [DHT](#distributed-hash-table-dht), and [CRUD actions](#crud-action) only add new data without removing old data.
+The property of a set of facts whereby the truth of prior facts are never negated by the addition of later facts. [CALM](#consistency-as-logical-monotonicity-calm-theorem) relies on functions that exhibit this property. For example, Holochain's [source chain](#source-chain), [DHT](#distributed-hash-table-dht) only add new data without removing old data, simulating deletions and modifications ([CRUD](#create-read-update-delete-crud)) by recording actions which override the status of, but don't remove, the data they refer to.
```
Comment (pdaoust) [inline src/pages/references/glossary.md:580]:
```suggestion
The property of a set of facts whereby the truth of prior facts are never negated by the addition of later facts. [CALM](#consistency-as-logical-monotonicity-calm-theorem) relies on functions that exhibit this property. For example, Holochain's [source chain](#source-chain) and [DHT](#distributed-hash-table-dht) only add new data without removing old data, simulating deletions and modifications ([CRUD](#create-read-update-delete-crud)) by recording actions which override the status of, but don't remove, the data they refer to.
```

File: src/pages/references/glossary.md [hunk 20]
```diff
@@ -404,95 +500,142 @@ See [Holochain host API](#holochain-host-api).
...
+
+<!--- removed because query arcs don't exist yet -- They also have a wider neighbourhood of nodes they can talk to to receive authoritative data, defined by their [query arc](#query-arc). -->
 
 #### Network
 
 In Holochain terms, a collection of [nodes](#node) [gossiping](#gossip) with each other to form a [validating DHT](#validating-dht), aiding in data storage and retrieval, [validation](#validation-rule), and peer discovery. Each [DNA](#dna) has a separate network.
 
+#### Network seed
+
+An optional string, specified in a [DNA bundle](#dna-bundle) file or passed at DNA installation time, that modifies the [DNA's hash](#dna-hash) without modifying any of its behavior. This can be used to create a unique [network](#network) shared by all [agents](#agent) who use the same network seed. Hence, a network seed is considered a [DNA modifier](#dna-modifiers).
```
Comment (pdaoust) [inline src/pages/references/glossary.md:637]:
```suggestion
An optional string, specified in a [DNA bundle](#dna-bundle) file or passed at [cell](#cell) [cloning](#cloning) time, that modifies the [DNA's hash](#dna-hash) without modifying any of its behavior. This can be used to create a unique [network](#network) shared by all [agents](#agent) who use the same network seed. Hence, a network seed is considered a [DNA modifier](#dna-modifiers).
```

File: src/pages/concepts/4_dht.md [hunk 21]
```diff
@@ -17,126 +17,145 @@ Agents share records of their actions, including any data meant to be shared wit
...
 
-* The **entry authority**, whose address is in the neighborhood of the entry’s address, checks the contents of the entry data to make sure it’s properly formed and conforms to the validation rules for the entry type. (Some records, like link creations and deletions, don’t have any entry data.)
-* The **action authority**, whose address is in the neighborhood of the action’s address, checks the contents of the action to make sure the previous action exists on the DHT and the action’s sequence ID and timestamp are higher than that of the previous action.
-* The **agent activity authority**, whose address is in the neighborhood of the record’s author (that is, they’re the author’s neighbors), receives a copy of the action and checks that it doesn’t conflict with another action — that is, the author isn’t trying to modify their history by publishing an alternate one.
+* A **store entry** operation, containing both the entry blob and the create-entry action header, goes to the **entry authorities** whose arcs cover the entry hash. When it's integrated, the authorities will hold the entry data at that base, along with metadata that contains the action header.
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/pages/concepts/4_dht.md:136]: marking as resolved so the conversation all happens at the first instance of this question.

**Comment by pdaoust** [comment]
@jost-s 

> I'm wondering in how far this body of work could/should be combined with the Holochain specs that are being added to the repo by @zippy

I definitely hear the desire to DRY it up -- I feel like they serve very different functions though; the glossary's purpose being a quick-reference for newcomers reading the literature as well as existing devs who've forgotten some things, whereas the other is more of a reference document for implementors and seasoned hApp devs. What are your thoughts on that?

**Comment by jost-s** [comment]
Why are you updating files in the archive by the way? They're not available on the site.

File: src/pages/concepts/8_calls_capabilities.md (hunk 15 above)
Comment (jost-s) [reply src/pages/concepts/8_calls_capabilities.md:31]: Yes, that is clearer to me. I'd drop "a default that helps protect the cell from being accessed by anyone other than its owner", because cells can be accessed by agents other than the owner.

File: src/pages/concepts/10_countersigning.md [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+
+If your application needs something more, you can build in strategies to discourage, prevent, or detect a double-spend attempt as it's happening. The simplest of these is to nominate an enzyme as a 'trusted notary', a witness whose job is to record agents' histories on its source chain and look for forks --- similar to agent activity authorities, but with the benefit of the determinism of a single source chain. Essentially, the enzyme becomes a consensus-maker.
+
+## M-of-N optional witnesses
+
+![](/assets/img/concepts/10.15-m-of-n-witnesses.png){.sz80p} {.center}
+
+Countersigning agents can also nominate **optional witnesses** to sign the transaction. This can be used to create a 'lightweight consensus', in which a majority of trusted witnesses must confirm, before signing, that no counterparty is attempting to fork their source chain. One of the witnesses must be an enzyme who collects all the required and optional signatures.
+
+While this requires more network chatter, it can tolerate more hardware and network failure than a single enzyme. Although an enzyme is still required, the enzyme role can be chosen at random or by rotation from among the pool of witnesses.
```
Comment (jost-s) [inline src/pages/concepts/10_countersigning.md:137]: without "from"?
```suggestion
While this requires more network chatter, it can tolerate more hardware and network failure than a single enzyme. Although an enzyme is still required, the enzyme role can be chosen at random or by rotation among the pool of witnesses.
```

**Review by jost-s — CHANGES_REQUESTED** [review]

File: src/pages/concepts/2_application_architecture.md (hunk 7 above)
Comment (pdaoust) [reply src/pages/concepts/2_application_architecture.md:106]: okay, I'll skip that change then. Thanks!

Comment (jost-s) [reply src/pages/concepts/2_application_architecture.md:106]: I stand correct after all. There is [an API call](https://docs.rs/holochain_conductor_api/latest/holochain_conductor_api/enum.AdminRequest.html#variant.UpdateCoordinators) which is functional AFAIK. What isn't there yet is a command to bundle the coordinators in the CLI.

**Comment by pdaoust** [comment]
Upgrading to an actual real PR, with more content improvements. @jost-s it still says you want me to make some changes; do you know if this is still true? (I know it's been a while)

**Review by jost-s — APPROVED** [review]

**Comment by jost-s** [comment]
> Upgrading to an actual real PR, with more content improvements. @jost-s it still says you want me to make some changes; do you know if this is still true? (I know it's been a while)

Nah, I just didn't do another review action after your had made the changes.
